### PR TITLE
Dask for parallel execution of the dag

### PR DESCRIPTION
**Pull request by comane**
This PR makes use of dask.distributed for a parallel evaluation of the dag generated by reportengine.ResourceBuilder.

### CODE REVIEW

**Comment by Zaharid** [comment]
Super cool :)

I'll add more detailed comments,but seems it is down to details like the handling of the plotting style.

File: src/reportengine/resourcebuilder.py [hunk 1]
```diff
@@ -188,72 +388,6 @@ def set_result(self, result, spec):
         for action, args in self._node_flags[spec]:
             action(result, self.rootns ,spec, **dict(args))
 
-
-
-    async def _run_parallel(self, deps, completed_spec):
-        try:
-            runnable_specs = deps.send(completed_spec)
-        except StopIteration:
```
Comment (comane) [inline src/reportengine/resourcebuilder.py:196]: @Zaharid, I deleted the old _run_parallel, however not so sure to which extent this is a good idea since at the moment the old method seems to deal better with large task as `vp-comparefits`

File: src/reportengine/figure.py [hunk 2]
```diff
@@ -60,18 +62,22 @@ def prepare_paths(*,spec, namespace, environment ,**kwargs):
...
     return {'paths':list(paths), 'output':environment.output_path}
 
+
+
 def savefig(fig, *, paths, output ,suffix=''):
     """Final action to save figures, with a nice filename"""
     #Import here to avoid problems with use()
     import matplotlib.pyplot as plt
+    import matplotlib
+    matplotlib.use('Agg')
```
Comment (Zaharid) [inline src/reportengine/figure.py:72]: This should not be needed any more.

**Comment by Zaharid** [comment]
As discussed, I'd prefer it if the `figuregen` decorator didn't modify the original function when used outside reportengine, and kept it a generator. Then again, that would be a fairly niche problem, so might do later.

**Review by Zaharid — CHANGES_REQUESTED** [review]
This seems to break the help system

```
/d/shm $ validphys --help plot_lumi1d                                (nnpdfdev) 
Traceback (most recent call last):
  File "/home/zah/anaconda3/envs/nnpdfdev/bin/validphys", line 33, in <module>
    sys.exit(load_entry_point('validphys', 'console_scripts', 'validphys')())
  File "/home/zah/nngit/nnpdf/validphys2/src/validphys/scripts/main.py", line 10, in main
    vp.main()
  File "/home/zah/nngit/nnpdf/validphys2/src/validphys/app.py", line 157, in main
    a.main()
  File "/home/zah/sourcecode/reportengine/src/reportengine/app.py", line 416, in main
    self.init()
  File "/home/zah/nngit/nnpdf/validphys2/src/validphys/app.py", line 124, in init
    super().init()
  File "/home/zah/sourcecode/reportengine/src/reportengine/app.py", line 319, in init
    args = self.get_commandline_arguments(cmdline)
  File "/home/zah/sourcecode/reportengine/src/reportengine/app.py", line 237, in get_commandline_arguments
    args = vars(self.argparser.parse_args(cmdline))
  File "/home/zah/anaconda3/envs/nnpdfdev/lib/python3.10/argparse.py", line 1826, in parse_args
    args, argv = self.parse_known_args(args, namespace)
  File "/home/zah/anaconda3/envs/nnpdfdev/lib/python3.10/argparse.py", line 1859, in parse_known_args
    namespace, args = self._parse_known_args(args, namespace)
  File "/home/zah/anaconda3/envs/nnpdfdev/lib/python3.10/argparse.py", line 2072, in _parse_known_args
    start_index = consume_optional(start_index)
  File "/home/zah/anaconda3/envs/nnpdfdev/lib/python3.10/argparse.py", line 2012, in consume_optional
    take_action(action, args, option_string)
  File "/home/zah/anaconda3/envs/nnpdfdev/lib/python3.10/argparse.py", line 1936, in take_action
    action(self, namespace, argument_values, option_string)
  File "/home/zah/sourcecode/reportengine/src/reportengine/app.py", line 55, in __call__
    self.app.init_style({})
  File "/home/zah/sourcecode/reportengine/src/reportengine/app.py", line 305, in init_style
    self.environment.default_style = self.default_style
AttributeError: 'App' object has no attribute 'environment'. Did you mean: 'make_environment'?
```

File: src/reportengine/resourcebuilder.py [hunk 3]
```diff
@@ -168,16 +212,187 @@ def execute_sequential(self):
...
+        scheduler : str, default: None
+                socket port number of dask scheduler.
+                A dask scheduler with associated dask workers should be initiated.
+                The socket port number should be passed to, e.g. valiphys, command line
+                when running it in --parallel mode.
+                
+        """
+        log.info("Initializing dask.distributed Client")
+        # set logger to CRITICAL
+        logging.getLogger("distributed").setLevel(logging.CRITICAL)
```
Comment (Zaharid) [inline src/reportengine/resourcebuilder.py:249]: Why is this needed? Does it do anything if the client is remote?

**Review by Zaharid — APPROVED** [review]

**Comment by Zaharid** [comment]
@comane I am fine with this now.
Feel free to merge!
